### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by mm-binary**

### CODE REVIEW

File: lib/VolSurface/Calibration/SABR.pm [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+
+requires qw(symbol surface term_by_day smile_points parameterization);
+
+use Date::Utility;
+use List::MoreUtils qw(pairwise first_index indexes);
+use List::Util qw(min max);
+use Math::Trig qw(tanh);
+use Try::Tiny;
+use Format::Util::Numbers qw(roundnear);
+#use DataDog::DogStatsd::Helper qw(stats_gauge);
```
Comment (jy-binary) [inline lib/VolSurface/Calibration/SABR.pm]: remember to remove this line

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    - "5.12"
+services:
+    - redis-server
+before_install:
+    - "cpanm JSON"
+    - "cpanm Date::Utility"
+    - "cpanm Format::Util::Numbers"
+    - "cpanm Test::FailWarnings"
+    - "cpanm Test::Most"
+    - "cpanm Devel::Cover::Report::Coveralls"
```
Comment (jy-binary) [inline .travis.yml]: I don't think we need coveralls

File: lib/VolSurface/Calibration/SABR.pm (hunk 1 above)
Comment (mm-binary) [inline lib/VolSurface/Calibration/SABR.pm]: Sure.

File: .travis.yml (hunk 2 above)
Comment (mm-binary) [inline .travis.yml]: I thought that it is needed for "-report coveralls" in the last line of the file. I will check and If not needed, I will delete it.

File: lib/VolSurface/Calibration/SABR.pm [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+    default => 1e-05,
+);
+
+has _scale => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 10,
+);
+
+=head2 function_to_optimize
```
Comment (tfoertsch) [inline lib/VolSurface/Calibration/SABR.pm]: you want an empty line after this one

File: lib/VolSurface/Calibration/SABR.pm [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+The function that we want to optimize.
```
Comment (tfoertsch) [inline lib/VolSurface/Calibration/SABR.pm]: and also after this one

File: Makefile.PL [hunk 5]
```diff
@@ -0,0 +1,26 @@
+use 5.006;
```
Comment (tfoertsch) [inline Makefile.PL]: do we really want this?

File: lib/VolSurface/Calibration/SABR.pm [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+    # this number is returned when the optimization is heading the wrong path
+    my $error = 1000;
+
+    #if (not $params) {
+    #    $self->_logger->logcroak('initial params are not present during optimization');
+    #}
+
+    my %params =
+        pairwise { our $a => our $b } @{$self->calibration_param_names},
+        @$params;
```
Comment (tfoertsch) [inline lib/VolSurface/Calibration/SABR.pm]: perhaps a bit more readable and without an external module:

```perl
my %params;
@params{@{$self->calibration_param_names}} = @$params;
```

File: lib/VolSurface/Calibration/SABR.pm [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+
+    my @tenors = @{$self->term_by_day};
+    my @kurtosis =
+        map { $self->_calculate_kurtosis($_ / 365, \%params) } @tenors;
+    my @variance =
+        map { $self->_calculate_variance($_ / 365, \%params) } @tenors;
+    my @skew = map { $self->_calculate_skew($_ / 365, \%params) } @tenors;
+
+    foreach my $param_name (keys %params) {
```
Comment (tfoertsch) [inline lib/VolSurface/Calibration/SABR.pm]: There is no order in the output of `keys`. To get a stable result it's possibly better to use `@{$self->calibration_param_names}` here.

**Comment by tfoertsch** [comment]
image

File: lib/VolSurface/Calibration/SABR.pm (hunk 3 above)
Comment (mm-binary) [inline lib/VolSurface/Calibration/SABR.pm]: Fixed (Also for other similar cases).

File: lib/VolSurface/Calibration/SABR.pm (hunk 4 above)
Comment (mm-binary) [inline lib/VolSurface/Calibration/SABR.pm]: Done.

File: Makefile.PL (hunk 5 above)
Comment (mm-binary) [inline Makefile.PL]: Changed to 5.014

**Pull request closed without merging**